experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,258 @@
+        let mut index3_pos: u32 = (code_point >> SHIFT_3) & INDEX_3_MASK;
+        let mut data_block: u32;
+        if index3_block & 0x8000 == 0 {
+            // 16-bit indexes
+            data_block = self
+                .index
+                .get((index3_block + index3_pos) as usize)
+                .unwrap() as u32;
+        } else {
+            // 18-bit indexes stored in groups of 9 entries per 8 indexes.
+            index3_block = (index3_block & 0x7fff) + (index3_pos & !7) + (index3_pos >> 3);
+            index3_pos &= 7;
+            data_block = ((self.index.get(index3_block as usize).unwrap() << (2 + (2 * index3_pos)))
+                as u32
+                & 0x30000) as u32;
+            index3_block += 1;
+            data_block |= self
+                .index
+                .get((index3_block + index3_pos) as usize)
+                .unwrap() as u32;
+        }
+        // Return data_pos == data_block (offset) +
+        //     portion of code_point bit field for last (4th) lookup
+        data_block + (code_point & SMALL_DATA_MASK)
+    }
+
+    /// Internal trie getter for a code point at or above the fast limit that
+    /// is designed for this trie's trie type, [`T`]. Returns the position in
+    /// the `data` array for the value that is associated with `code_point`.
+    fn trie_above_fastmax_index(&self, code_point: u32) -> u32 {
+        if code_point >= self.header.high_start {
+            self.header.data_length - HIGH_VALUE_NEG_DATA_OFFSET
+        } else {
+            self.trie_internal_small_index(code_point) // helper fn
+        }
+    }
+
+    /// Internal trie getter for a code point below the fast limit that
+    /// is designed for this trie's trie type, [`T`]. Returns the position in
+    /// the `data` array for the value that is associated with `code_point`.
+    fn trie_below_fastmax_index(&self, c: u32) -> u32 {
+        let index_array_pos: u32 = c >> FAST_TYPE_SHIFT;
+        let index_array_val: u16 = self.index.get(index_array_pos as usize).unwrap(); // 1. How to specify type parameter for .index() 2. How to avoid unwrap()?
+        let fast_index_val: u32 = index_array_val as u32 + (c & FAST_TYPE_DATA_MASK);
+        fast_index_val
+    }
+
+    /// Return the value that is associated with `code_point` for this [`CodePointTrie`].
+    pub fn get(&self, code_point: u32) -> W {
+        let data_pos: u32 = if code_point <= T::FAST_MAX {
+            Self::trie_below_fastmax_index(self, code_point)
+        } else if code_point <= CODE_POINT_MAX {
+            Self::trie_above_fastmax_index(self, code_point)
+        } else {
+            self.header.data_length - ERROR_VALUE_NEG_DATA_OFFSET
+        };
+        // need the unwrap because the default value is stored in the data array,
+        // and getting that default value always returns an Option<W>, but need to return W
+        self.data.get(data_pos as usize).unwrap()
```

Comment:
It sounds like the default value is stored differently from other values. Why would that be?

Reply:
Let me know what you think, but I think this is a difference in idiomatic array lookup between C/C++ and Rust. 

IIUC, in C/C++, you always get back a value when you index into an array, and one way of indicating an error is to return a sentinel value, which is purpose of the errorValue that is located at the errorIndex of the data array. The tricky thing in attempting to port that behavior to Rust is that Rust [lookups in a vector](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get) always returns an `Option<T>` type. The API for [ZeroVec for get](https://unicode-org.github.io/icu4x-docs/doc/zerovec/enum.ZeroVec.html#method.get) was designed consistently.

So with the ZeroVec API that return an `Option<T>`, if it has a problem, it returns a `None` rather than `Some(my_value)`, and then to return the errorValue in the trie's data array, I would have to do another lookup to get the errorValue returned as an `Option<T>` type and call `.unwrap()` again on the faith that the lookup was successful and provided `Some(errorValue)`. So it seemed redundant. (`.unwrap()` will throw an exception, aka "panic", if the result is `None`).

But as I think about this as I type this, given the faith required to port the code with all of the `.unwrap()` calls, maybe it can't hurt to go ahead and add the extra lookup that returns the errorValue ?

Reply:
In the vast majority of cases, it should be possible to write code without `.unwrap()`.

Could you us some other function that puts a replacement value in for the missing value, instead of panicking?  For example, `unwrap_or_else()` or `unwrap_or_default()`

Reply:
FWIW, in a correctly built trie the array indexing will always be valid. It's ok to panic only if there is garbage.